feat: Add "certificate_available_date" in edX's Course Details API

aliraza-abbasi opened the pull request:
<!--

🎉🎉 Olive has been released! 🎉🎉

🫒🫒
🫒🫒🫒🫒         🫒 Note: Olive is in support. Fixes you make on master may still
    🫒🫒🫒🫒     be needed on Olive. If so, make another pull request against the
🫒🫒🫒🫒         open-release/olive.master branch, or ping @nedbat for help or questions.
🫒🫒

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

We have a feature in edX called Certificate Display Behaviour that manages how the certificates are made available to learners other than the course's end date. The problem is that edX doesn't have this key in its course detail API or any other API that we can use independent of enrollment in the course. To enable this behavior in edX Add the `certificate_available_date` in the course details API.

Useful information to include:
- This PR will only add a field `certificate_available_date` in Course Detail [API](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_read) `GET /api/courses/v1/courses/{course_key}/`.

## Supporting information

The related issue is https://github.com/mitodl/mitxonline/issues/1339.

## Testing instructions

- Checkout to this branch
- Run the edx app
- Hit the course detail API on your local server ([Ref](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_read))
- Verify that `certificate_available_date` is in the response of this API

## Deadline

None

## Other information

None

openedx-webhooks commented:
Thanks for the pull request, @aliraza-abbasi!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pdpinch commented on docs/swagger.yaml:
```diff
@@ -2196,6 +2196,7 @@ paths:
                 * `"timestamp"`: generated from the `start` timestamp
                 * `"empty"`: no start date is specified
             * pacing: Course pacing. Possible values: instructor, self
+            * certificate_available_date: Date the certificate availble, in ISO 8601 notation
```
```suggestion
            * certificate_available_date: Date the certificate will be available, in ISO 8601 notation
```

pdpinch commented on docs/swagger.yaml:
```diff
@@ -2240,7 +2241,8 @@ paths:
                     "start_type": "timestamp",
-                    "pacing": "instructor"
+                    "pacing": "instructor",
+                    "certificate_available_date": "2015-08-14T00:00:00Z"
```
Can we get away with this?

```suggestion
                    "certificate_available_date": "2015-08-14T00:00:00Z", 
```

aliraza-abbasi replied:
`,` at the last row, I think is invalid in JSON format and also it was not present previously that's why I've not added this at the last line

pdpinch commented:
nit: I think the commit message should be a "feature" instead of a "fix"

aliraza-abbasi committed:
feat: Enable Certificate Display Behavior add certificate_available_date in course detail API

feat: Enable Certificate Display Behavior add certificate_available_date in course detail API

fix: reviwed changes, certificate_available_date in Course Detail API but not in Course List API

fix: reviewed changes, certificate available date display condition updated

fix: final reviewed changes

fix: serializer and linting tests

fix: serializer and tests

fix: docstring

fix: docstring and tests

fix: typo and mock call

arslanashraf7 commented on lms/djangoapps/course_api/serializers.py:
```diff
@@ -113,6 +113,7 @@ class CourseSerializer(serializers.Serializer):  # pylint: disable=abstract-meth
     mobile_available = serializers.BooleanField()
     hidden = serializers.SerializerMethodField()
     invitation_only = serializers.BooleanField()
+    certificate_available_date = serializers.DateTimeField()
```
I think we can improve this a little bit, Currently `certificate_available_date` is returned no matter if the certificate display behavior is enabled or not. Although this might be more like a business decision I'd mention what I think here should be done here:

1. I think the addition of this field here in the course details API should be conditional on whether to include this date in the API or not.
2. The conditions that I propose are that before adding this to the response of course details API we should check if the certificate display behavior is enabled or not. 
3. We should only include `certificate_available_date` in the API if the `certificate_display_behaviour` is enabled otherwise this can be None.

For point#3, _How do we enable this behavior?_ I've mentioned it in https://github.com/mitodl/mitxonline/issues/1339#issue-1530615721.

As a reference, I looked at the implementation of other APIs (e.g. [Courseware API](https://courses.edx.org/api-docs/#/courseware/courseware_course_read)) that include this date and they also check for the certificate display behavior to be enabled before adding the `certificate_available_date` into the API response.

[Here](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/certificates/api.py#L299-L311) is how it's done for the courseware API in the code.

@pdpinch What are your thoughts on this?

pdpinch replied:
I agree, particularly because the other APIs check this course setting as well.

aliraza-abbasi replied:
@pdpinch Do we need `certificate_available_date` in the [CourseListAPI](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_list) and [CourseDetailAPI](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_read) or only in [CourseDetailAPI](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_read)?

pdpinch replied:
I don't know how [CourseListAPI](https://courses.edx.org/api-docs/#/courses/courses_v1_courses_list) is used, but I would prefer that we don't expand the scope of this PR.

arslanashraf7 reviewed (changes requested):
I've added a suggestion.

arslanashraf7 commented on lms/djangoapps/course_api/api.py:
```diff
@@ -79,6 +83,24 @@ def course_detail(request, username, course_key):
+            overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE
+        )
+    else:
+        display_behavior_is_valid = True
```
I think we should convert this into a common function instead of copying the code itself. The function can then be used in different places like the one I mentioned above in the reference in my last comment. We'll also have to refactor the existing usages of this code, I expect there would not be so many places that will need refactoring.

Although, I agree with @pdpinch that we should not extend the scope of the PR this is something that we should do to avoid unintended problems in the future.

arslanashraf7 commented on lms/djangoapps/course_api/tests/test_serializers.py:
```diff
@@ -164,6 +164,7 @@ def setUp(self):
         # update the expected_data to include the 'overview' data.
         about_descriptor = XBlock.load_class('about')
         overview_template = about_descriptor.get_template('overview.yaml')
+        self.expected_data['certificate_available_date'] = '2015-08-14T00:00:00Z'
```
I think we should also add a test where this is `None` when the certificate display behavior is not turned on.

aliraza-abbasi replied:
This check is not on the `serializers.py` so I can't add it in `test_serializers.py` because the serializer will return according to the received value if it's `None` it'll return `None` else it'll return a `DateTime` value. I've added the `ENABLE_V2_CERT_DISPLAY_SETTINGS` check in [api.py:course_detail](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_api/api.py#L56) So the test can be added here in [test_api.py:TestGetCourseDetail](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_api/tests/test_api.py#L65). But I've explored the tests here and I've not found any test case to check response fields. I think we don't need a test case for `None`

@pdpinch What are your thoughts on this? As @arslanashraf7 is on leaves

arslanashraf7 reviewed (changes requested):
I left a couple of suggestions, mostly related to the DRY approach and refactoring.

rhysyngsun commented on lms/djangoapps/course_api/api.py:
```diff
@@ -79,6 +81,13 @@ def course_detail(request, username, course_key):
+        return overview
+    current_status = certificate_status_for_student(user, course_key)
+
+    if not allow_certificate_available_date(overview, current_status):
```
I think because this is not a user-specific API that we shouldn't be conditionally displaying this based on the state of a student's certificate. I think this check should actually be what `allow_certificate_available_date` is, but without `CertificateStatuses.is_passing_status(current_status['status'])`.

pdpinch replied:
I'm not following this completely, but we should avoid mixing together code that returns course metadata with code that returns student data.

arslanashraf7 replied:
It's a good point by @rhysyngsun. I agree we should not check student states for `certificate_available_date` in the API.

rhysyngsun commented on lms/djangoapps/certificates/api.py:
```diff
@@ -268,6 +268,38 @@ def generate_certificate_task(user, course_key, generation_mode=None):
+    """
+    certificate_status = (
+        CertificateStatuses.is_passing_status(certificate_status['status'])
+        if certificate_status
```
```suggestion
        if certificate_status is not None
```

arslanashraf7 commented on lms/djangoapps/certificates/api.py:
```diff
@@ -268,6 +268,38 @@ def generate_certificate_task(user, course_key, generation_mode=None):
+        certificate_status and
+        display_behavior_is_valid and
+        course_overview.certificate_available_date
+    )
```
This might be a long comment but I was digging around to see how generic we can make this and use this commonly with reference to my previous comment about a reusable function to check the display behavior setting here and in other places.

I was looking at different APIs that return `certificate_available_date` 
So, while looking around I saw many usages of checking the display behavior in detail. These APIs are mentioned in https://github.com/mitodl/mitxonline/issues/1257#issuecomment-1346463078 so I will try not to re-write all that.

In my analysis, i noticed that these APIs at different points call https://github.com/openedx/edx-platform/blob/96b42271e3041f9ce4cd4a53566b7907569e902c/lms/djangoapps/courseware/views/views.py#L1070 in combination with different checks on certificate status. This function internally calls other functions that deeply involve student and certificate status which i really think we don't need.

e.g. in the selected lines here if you check  `not certificates_viewable_for_course(course_overview)` in detail it also checks different types of certificate state checks that we might not need.

Here are my testing results:
1. This worked partially I think, e.g. if disabling the waffle switch `certificates.auto_certificate_generation` which turns off the display behavior setting in the studio, I still get the `certificate_available_date` in the course details API although the behavior is disabled in the studio.
2. The above point made me more curious about digging into the APIs in detail that returns `cert_data` in response and I think if we were to create this utility function and use it everywhere we'll have to make a lot of changes in different pieces of code that will increase the scope of this and we might end up breaking something that we don't intend to.

**My suggestion:**

I think for this specific API, we should only check for `ENABLE_V2_CERT_DISPLAY_SETTINGS` and `certificates.auto_certificate_generation`. These are the exact setting that enables certificate display behavior in studio and these should also enable the addition of `certificate_available_date` in the course details API.

What i think is that we should remove this function and do something like:

```
if settings.FEATURES.get("ENABLE_V2_CERT_DISPLAY_SETTINGS") and auto_certificate_generation_enabled():
        display_behavior_is_valid = (
            course_overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE
        )
    else:
        display_behavior_is_valid = True

// If the above is True, add the `certificate_available_date` in the course details API. 
```

and undo the current refactoring we did below in `certificate_downloadable_status`. This would make the PR simpler, touching the least parts of the other functionality.

rhysyngsun replied:
I think your suggestion makes sense to me.

aliraza-abbasi replied:
I’ve already removed this. Please review it again

rhysyngsun commented:
@aliraza-abbasi looks good to me

arslanashraf7 commented on lms/djangoapps/course_api/api.py:
```diff
@@ -79,6 +81,9 @@ def course_detail(request, username, course_key):
     overview.effective_user = user
+    if settings.FEATURES.get("ENABLE_V2_CERT_DISPLAY_SETTINGS") and auto_certificate_generation_enabled():
+        if not overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE:
+            overview.certificate_available_date = None
```
I think this might be the last comment before I could approve the PR. I took the liberty of looking at exactly how this has been getting enabled in Studio and [this is how](https://github.com/openedx/edx-platform/blob/master/cms/templates/settings.html#L255-L299) is being enabled in Studio course dates page.

So what it does is, it checks if the waffle switch for certificates auto-generation is enabled it checks if a course is self-paced or instructor paced, If the course is instructor paced it enables the behavior right away that looks something like:
<img width="899" alt="image" src="https://user-images.githubusercontent.com/34372316/220099568-33c5377e-6afc-4a01-bf32-5f2026ea9a2b.png">

So it only checks the v2 settings as mentioned in the ticket https://github.com/mitodl/mitxonline/issues/1339 only to display multiple date options as seen in 
<img width="886" alt="image" src="https://user-images.githubusercontent.com/34372316/210965487-e4b63049-954f-457d-96ab-23a15f2dcb3b.png">
.

So I think we can make one last iteration and make it even simpler but simply calling https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/certificates/api.py#L840 and adding the certificate available date to the API else None.

@aliraza-abbasi Let me know if it makes sense to you.

aliraza-abbasi replied:
Thank you @arslanashraf7. I think this would be much better. It makes sense to me as well. I'll update the code

arslanashraf7 reviewed (changes requested):
Just the last changes.

arslanashraf7 commented:
@aliraza-abbasi Before i review this, Could you do a rebase on master if not already?

aliraza-abbasi commented:
> @aliraza-abbasi Before i review this, Could you do a rebase on master if not already?

<strike>I had already rebased, My last push was to rebase, I'll rebase again if it gets approved and needs to be merged</strike>

I've rebased it again

rhysyngsun commented:
@aliraza-abbasi @arslanashraf7 generally this looks good to me, I'll leave it to you guys to move it forward with the maintainers.

arslanashraf7 commented on docs/swagger.yaml:
```diff
@@ -2196,6 +2196,8 @@ paths:
                 * `"timestamp"`: generated from the `start` timestamp
                 * `"empty"`: no start date is specified
             * pacing: Course pacing. Possible values: instructor, self
+            * certificate_available_date (optional): Date the certificate will be availble, in ISO 8601 notation
```
typo

```suggestion
            * certificate_available_date (optional): Date the certificate will be available, in ISO 8601 notation
```

arslanashraf7 commented on lms/djangoapps/course_api/serializers.py:
```diff
@@ -159,6 +160,16 @@ def get_overview(self, course_overview):
+    def to_representation(self, instance):
+        """
+        Get the `certificate_available_date` in response
+        if `auto_certificate_generation` is enabled
```
This will also help in refactoring in future if the switch name changes.

```suggestion
        if the `certificates.auto_certificate_generation` switch is enabled
```

arslanashraf7 commented on lms/djangoapps/course_api/tests/test_serializers.py:
```diff
@@ -164,8 +165,19 @@ def setUp(self):
 
+    def test_basic(self):
+        course = self.create_course()
+        CourseDetails.update_about_video(course, 'test_youtube_id', self.staff_user.id)
```
This is not needed here.

arslanashraf7 commented on lms/djangoapps/course_api/tests/test_serializers.py:
```diff
@@ -164,8 +165,19 @@ def setUp(self):
+        course = self.create_course()
+        CourseDetails.update_about_video(course, 'test_youtube_id', self.staff_user.id)
+        course_overview = CourseOverview.get_from_id(course.id)
+        if can_show_certificate_available_date_field(course_overview):
```
This test needs to run two times with params. Once with the switch on and once with the off to test the whole logic.

arslanashraf7 commented on lms/djangoapps/course_api/tests/test_serializers.py:
```diff
@@ -164,8 +165,19 @@ def setUp(self):
+
         self.expected_data['overview'] = overview_template.get('data')
 
+    def test_basic(self):
```
This name needs to be according to what this is testing. An we should also add docstrings

aliraza-abbasi replied:
I've to override this function from the parent class. Can't change its name

aliraza-abbasi replied:
There is no docstring in any of the above tests, So I left it as like others

arslanashraf7 commented on lms/djangoapps/course_api/views.py:
```diff
@@ -59,6 +59,8 @@ class CourseDetailView(DeveloperErrorViewMixin, RetrieveAPIView):
             * `"empty"`: no start date is specified
         * pacing: Course pacing. Possible values: instructor, self
+        * certificate_available_date (optional): Date the certificate will be availble, in ISO 8601 notation
+            if `auto_certificate_generation` is enabled
```
```suggestion
            if the `certificates.auto_certificate_generation` waffle switch is enabled
```

arslanashraf7 reviewed (approved):
Looks good 👍

openedx-webhooks commented:
@aliraza-abbasi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/31588&entry.1484655318=https://github.com/aliraza-abbasi&entry.752974735=2023-01-18+09:02&entry.1917517419=2023-02-24+02:04&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
